download/src/lib.rs

```diff
@@ -61,74 +61,81 @@ pub fn download_to_path_with_backend(
     use std::fs::OpenOptions;
     use std::io::{Read, Seek, SeekFrom, Write};
 
-    || -> Result<()> {
-        let (file, resume_from) = if resume_from_partial {
-            let possible_partial = OpenOptions::new().read(true).open(path);
-
-            let downloaded_so_far = if let Ok(mut partial) = possible_partial {
-                if let Some(cb) = callback {
-                    cb(Event::ResumingPartialDownload)?;
-
-                    let mut buf = vec![0; 32768];
-                    let mut downloaded_so_far = 0;
-                    loop {
-                        let n = partial.read(&mut buf)?;
-                        downloaded_so_far += n as u64;
-                        if n == 0 {
-                            break;
+    {
+        async move {
+            let (file, resume_from) = if resume_from_partial {
+                // TODO: blocking call
+                let possible_partial = OpenOptions::new().read(true).open(path);
+
+                let downloaded_so_far = if let Ok(mut partial) = possible_partial {
+                    if let Some(cb) = callback {
+                        cb(Event::ResumingPartialDownload)?;
+
+                        let mut buf = vec![0; 32768];
+                        let mut downloaded_so_far = 0;
+                        loop {
+                            let n = partial.read(&mut buf)?;
+                            downloaded_so_far += n as u64;
+                            if n == 0 {
+                                break;
+                            }
+                            cb(Event::DownloadDataReceived(&buf[..n]))?;
                         }
-                        cb(Event::DownloadDataReceived(&buf[..n]))?;
-                    }
 
-                    downloaded_so_far
+                        downloaded_so_far
+                    } else {
+                        let file_info = partial.metadata()?;
+                        file_info.len()
+                    }
                 } else {
-                    let file_info = partial.metadata()?;
-                    file_info.len()
-                }
-            } else {
-                0
-            };
-
-            let mut possible_partial = OpenOptions::new()
-                .write(true)
-                .create(true)
-                .open(path)
-                .context("error opening file for download")?;
+                    0
+                };
 
-            possible_partial.seek(SeekFrom::End(0))?;
-
-            (possible_partial, downloaded_so_far)
-        } else {
-            (
-                OpenOptions::new()
+                // TODO: blocking call
```

Comment:
Should this be addressed before merging?

Reply:
oh, I missed that I had these TODOs left. I will review.

Reply:
I think this is fine a as blocking call for now : we download a small number of files during install, and it is very unlikely that this is going to cause a performance problem. Eventually of course we should do it, but the goal here was to get a tokio runtime in place at parity, not to make perf/concurrency improvements, since those will require reasoning about resource limits and so on in a very careful fashion.